pkg/aqua/aqua.go

```diff
@@ -0,0 +1,237 @@
+		All:            true,
+		AQUAVersion:    AQUA_VERSION, // we need to regularly check and update this version
+	}
+
+	ctrl, err := controller.New(ctx, param)
+	errz.Fatal(err)
+
+	// Finally call installation
+	err = ctrl.Install(context.Background(), param)
+	errz.Fatal(err)
+
+	return nil
+}
+
+// Prune - remove all packages and undelying structures
+func (d *Definition) Prune(ctx context.Context) (err error) {
+	defer errz.Recover(&err)
+
+	aquaRoot, err := filepath.Abs(AQUA_ROOT)
+	errz.Fatal(err)
+
+	err = os.RemoveAll(aquaRoot)
+	errz.Fatal(err)
+
+	return nil
+}
+
+// Setup run/build environment so that packages can be accessed
+func (d *Definition) SetEnvirionment() error {
+	// Will hide internally used paths
+	_, err := d.setEnvirionment()
+	return err
+}
+
+func (d *Definition) Search(ctx context.Context) (pckgs []string, err error) {
+	defer errz.Recover(&err)
+	// Create local .aqua dir and reference to that
+	err = os.MkdirAll(AQUA_ROOT, os.ModePerm)
+	errz.Fatal(err)
+
+	// Setup envirionment
+	aquaEnv, err := d.setEnvirionment()
+	errz.Fatal(err)
+
+	// Create aqua.yaml file used inside aqua controller install call
+	err = d.createDefinitionFile(aquaEnv.AquaConfig)
+	errz.Fatal(err)
+
+	param := &controller.Param{
+		ConfigFilePath: aquaEnv.AquaConfig, // This should be nested somewhere inside .bob dir
+		IsTest:         false,
+		All:            true,
+		AQUAVersion:    AQUA_VERSION, // we need to regularly check and update this version
+	}
+
+	ctrl, err := controller.New(ctx, param)
+
+	errz.Fatal(err)
+
+	// store old stdout
+	old := ctrl.Stdout
+
+	r, w, err := os.Pipe()
+	errz.Fatal(err)
+
+	ctrl.Stdout = w
+
+	outC := make(chan string)
+	// copy the output in a separate goroutine so printing can't block indefinitely
+	go func() {
+		var buf bytes.Buffer
+		_, err := io.Copy(&buf, r)
+		errz.Fatal(err)
+		outC <- buf.String()
+	}()
+
+	err = ctrl.Generate(ctx, param)
+	errz.Fatal(err)
+
+	// back to normal state
+	err = w.Close()
+	errz.Fatal(err)
+	ctrl.Stdout = old // restoring the real stdout
+	out := <-outC
+
+	// reading our temp stdout
+	pckgs = strings.Split(out, "\n")
+
+	// remove AQUA_PACKAGE_PREFIX to give a clean package name and version string to the caller
+	for idx := range pckgs {
+		pckgs[idx] = strings.TrimPrefix(pckgs[idx], AQUA_PACKAGE_PREFIX)
+	}
+
+	return pckgs, nil
+}
+
+func (d *Definition) setEnvirionment() (env EnvionmentVariables, err error) {
+	defer errz.Recover(&err)
+
+	// Add aqua root to env
+	aquaRoot, err := filepath.Abs(AQUA_ROOT)
+	errz.Fatal(err)
+
+	err = os.Setenv(ENV_AQUA_ROOT_DIR, aquaRoot)
+	errz.Fatal(err)
+
+	// Add global aqua config file to env
+	aquaConfig, err := filepath.Abs(AQUA_FILE_PATH)
+	errz.Fatal(err)
+
+	err = os.Setenv(ENV_AQUA_GLOBAL_CONFIG, aquaConfig)
+
+	// Add aqua bin to PATH
+	aquaBin, err := filepath.Abs(fmt.Sprintf("%s/bin", aquaRoot))
+	errz.Fatal(err)
+
+	err = os.Setenv(ENV_PATH, fmt.Sprintf("%s:%s", aquaBin, os.Getenv(ENV_PATH)))
```

Comment:
This only works cause we pass the complete environment to the built-in shell of each build task. In the future we will  reconsider this due env-vars impacting the build outcome without being included in the input hash.

I'd recommend to set the aqua path explicitly on each build task, see here https://github.com/benchkram/bob/blob/main/bobtask/run.go#L31